events/api.py

```diff
@@ -1555,37 +1562,78 @@ def filter_queryset(self, queryset):
         """
         TODO: convert to use proper filter framework
         """
-        self.data_source, self.organization = get_authenticated_data_source_and_publisher(self.request)
         queryset = super(EventViewSet, self).filter_queryset(queryset)
-        auth_filters = Q(publication_status=PublicationStatus.PUBLIC)
-        if self.organization:
-            # USER IS AUTHENTICATED
-            if 'show_all' in self.request.query_params:
-                # Show all events for this organization,
-                # along with public events for others.
-                auth_filters |= get_publisher_query(self.organization)
-        queryset = queryset.filter(auth_filters)
+
+        if self.request.method in SAFE_METHODS:
+            auth_filters = Q(publication_status=PublicationStatus.PUBLIC)
+            if self.organization:
+                # USER IS AUTHENTICATED
+                if 'show_all' in self.request.query_params:
+                    # Show all events for this organization,
+                    # along with public events for others.
+                    auth_filters |= get_publisher_query(self.organization)
+            queryset = queryset.filter(auth_filters)
+        else:
+            # prevent changing events user does not have write permissions (for bulk operations)
+            queryset = self.request.user.get_editable_events(self.organization, queryset)
+
         queryset = _filter_event_queryset(queryset, self.request.query_params,
                                           srs=self.srs)
         return queryset.filter()
```

Comment:
So what is the exception that the API will return for POST and PUT if the event is not in the queryset after filtering, i.e. the user has no edit rights? I.e. does the API return HTTP 401, 403 or 404?

Reply:
Ah, apparently the tests specify it is 403, that's good. I just don't get where the logic for showing 403 is hidden after filtering has been done :D

Reply:
That's a very good question. In general, if a object is neither readable nor editable by a user, then it's better to return a 404 for all requests, but if the object is readable but not editable by user, then it should return 403 when user tries to edit the object.

Normally we want the `get_object` method to call the `filter_queryset` method to filter out unwanted objects first and then tries to get the wanted object from the filtered queryset, so it's not a good  idea to return different querysets for read requests (GET) and write requests (PUT, POST) for `filter_queryset` method as it means we can get 200 for GET but 404 for PUT and POST on the same object URL.

In our code, we are using a custom `get_object` method which does not use `filter_queryset` method at all, which means the `filter_queryset` is only used multi-objects operations (list and bulk updates). Thus we do not have the problem above.

Another options if to put this logic to the `perform_update` method, but we will write more lines of code. Cannot say which is better, just choose one that I thought simpler :smile: